packages/cli/generators/controller/index.js

```diff
@@ -34,25 +65,151 @@ module.exports = class ControllerGenerator extends ArtifactGenerator {
+    return this.prompt([
+      {
+        type: 'list',
+        name: 'controllerType',
+        message: 'What kind of controller would you like to generate?',
+        when: this.artifactInfo.controllerType === undefined,
+        choices: [ControllerGenerator.BASIC, ControllerGenerator.CRUD],
+        default: ControllerGenerator.BASIC,
+      },
+    ])
+      .then(props => {
+        Object.assign(this.artifactInfo, props);
+        return props;
+      })
+      .catch(err => {
+        debug(`Error during controller type prompt: ${err.stack}`);
+        return this.exit(err);
+      });
+  }
+
+  promptArtifactCrudVars() {
+    let modelList = [];
+    let repositoryList = [];
+    if (
+      !this.artifactInfo.controllerType ||
+      this.artifactInfo.controllerType === ControllerGenerator.BASIC
+    ) {
+      return;
+    }
+    return utils
+      .getArtifactList(this.artifactInfo.modelDir, 'model')
+      .then(list => {
+        if (_.isEmpty(list)) {
+          return this.exit(
+            new Error(`No models found in ${this.artifactInfo.modelDir}`)
+          );
+        }
+        modelList = list;
+        return utils.getArtifactList(
+          this.artifactInfo.repositoryDir,
+          'repository',
+          true
+        );
+      })
+      .then(list => {
+        if (_.isEmpty(list)) {
+          return this.exit(
+            new Error(
+              `No repositories found in ${this.artifactInfo.repositoryDir}`
+            )
+          );
+        }
+        repositoryList = list;
+        return this.prompt([
+          {
+            type: 'list',
+            name: 'modelName',
+            message:
+              'What is the name of the model to use with this CRUD repository?',
+            choices: modelList,
+            when: this.artifactInfo.modelName === undefined,
+            default: modelList[0],
+          },
+          {
+            type: 'list',
+            name: 'repositoryName',
+            message: 'What is the name of your CRUD repository?',
+            choices: repositoryList,
+            when: this.artifactInfo.repositoryName === undefined,
+            default: repositoryList[0],
+          },
+          {
+            type: 'list',
+            name: 'id',
+            message: 'What is the type of your ID?',
+            choices: ['number', 'string', 'object'],
+            when: this.artifactInfo.id === undefined,
+            default: 'number',
+          },
+        ]).then(props => {
+          debug(`props: ${inspect(props)}`);
+          Object.assign(this.artifactInfo, props);
+          // Ensure that the artifact names are valid.
+          [
+            this.artifactInfo.name,
+            this.artifactInfo.modelName,
+            this.artifactInfo.repositoryName,
+          ].forEach(item => {
+            const validationMsg = utils.validateClassName(item);
+            if (typeof validationMsg === 'string')
+              throw new Error(validationMsg);
```

Comment:
We should set up validate function for each of the prompts instead of checking after all answers are accepted. The prompt level validator will reject invalid inputs before the next one. See an example at https://github.com/strongloop/loopback-next/blob/master/packages/cli/generators/app/index.js#L45.